### DISCUSSION TITLE
ReportLab now supports PDF gradients, can we add that to svglib?

### DISCUSSION THREAD

**Discussion by scoofy** [deeplook/svglib#305] [Ideas]
I'm working on a project where i'm trying to add lots of SVGs to PDFs. Unfortunately linear and radial gradients will be a big part of that.

I noticed on this github entry that gradients have been available since 2013: https://stackoverflow.com/questions/452074/creating-a-gradient-fill-in-a-pdf-file-using-reportlab

Was thinking about trying to add that to the codebase, but i'll have to look through it. I don't know if anyone else is interested in helping, but I could definitely use the help understanding the codebase.

**Comment by claudep** [comment]
I think you should read #142, it looks like we are far from being able to support that, unfortunately :-(

**Reply by scoofy** [reply]
So I read that, and looked through the code as well. I'm new to the codebase, fully admitted, but since reportlab has gradients, i think it should be possible if return a series PDFAxialShading dictionaries *with* the drawing that's returned, and from there, it should be pretty easy to render those to the canvas on runtime with a wrapper. It would be tricky, but you should be able to grab the shading data exactly in the code where the convertColor function hits the "url(linearGradientXXXX)" string where the color fails.

The obvious drawback is that you wouldn't be able to simply return a drawing. The SVG rendering would need to go straight to a canvas on runtime, however, if would be pretty easy to write in a warning that lets people know if they want gradient rendering, they need to run the wrapper straight to a canvas. If that's a waste of everyone's time, i understand. The easiest solution for me is to just convert to PNG and treat it as an image, but i definitely wanted to help with the project if I could.

**Reply by deeplook** [reply]
@claudep is right. Even if the PDF canvas provides sufficient functionality for creating linear/axial gradients it is of no immediate use for svglib, as it operates on drawings, only. A wild idea that comes up to my mind is to add gradient primitives to svglib, store them in some kind of stack when parsing an SVG file, ignore them when generating the drawing, render the drawing on a PDF canvas, and "play back" the "recorded" gradients to that canvas. It would be some challenge and it might not play well with the expected order of operations. So I'm really not sure it's worth trying for you.

**Reply by scoofy** [reply]
Thanks for the response. Yes, the other idea that i had was literally to modify the canvas class drawing function to check for the saved gradients, but i realize that's pretty insane, though i doubt that reportlab is going to change much going forward. 

I was unable to find a PDF solution, even for converting to png/jpeg, with Wand/Cairo/etc or anything short of running an actual version of inkscape on my server and making command calls. I was able to find a reasonable solution for my specific issue by rendering printable html/css rather than convert my vectors to formatted PDF. It was an unfortunate choice (and i'd vastly prefer PDF), but it was doable in a short period of time. 

If you're interested in trying something wild, i'd absolutely be willing to help, by my xml-type knowledge is limited to an xbrl-to-json project i made a few years ago. I still think it's doable, and i'd absolutely be willing to put in some time.

**Reply by claudep** [reply]
In my opinion, it would be more generally profitable to invest time for adding pattern support in ReportLab itself.
